Sources/TuistCore/Graph/GraphTraverser.swift

```diff
@@ -266,27 +266,71 @@ public class GraphTraverser: GraphTraversing {
             test: isDependencyPrecompiledDynamicAndLinkable,
             skip: canDependencyEmbedProducts
         )
-        // Skip merged precompiled libraries from merging into the runnable binary
-        if case let .manual(dependenciesToMerge) = target.target.mergedBinaryType {
-            precompiledFrameworks = precompiledFrameworks
-                .filter { !isXCFrameworkMerged(dependency: $0, expectedMergedBinaries: dependenciesToMerge) }
-        }
-        references.formUnion(precompiledFrameworks.lazy.compactMap { self.dependencyReference(
-            to: $0,
-            from: .target(name: name, path: path)
-        ) })
-
         /// Other targets' frameworks.
         var otherTargetFrameworks = filterDependencies(
             from: .target(name: name, path: path),
             test: isDependencyDynamicTarget,
             skip: canDependencyEmbedProducts
         )
 
+        // Skip merged precompiled libraries from merging into the runnable binary first
+        switch target.target.mergedBinaryType {
+        case .automatic:
+            precompiledFrameworks = precompiledFrameworks
+                .filter { !isXCFrameworkMerged(dependency: $0) }
+        case let .manual(dependenciesToMerge):
+            precompiledFrameworks = precompiledFrameworks
+                .filter { !isXCFrameworkMerged(dependency: $0, expectedMergedBinaries: dependenciesToMerge) }
+        case .disabled:
+            break
+        }
+
+        // If runnable binary is a merged binary, keep only non-mergeable targets
         if target.target.mergedBinaryType != .disabled {
             otherTargetFrameworks = otherTargetFrameworks.filter(isDependencyDynamicNonMergeableTarget)
         }
 
+        // Search for merged non-runnable binaries
+        for innerTarget in otherTargetFrameworks {
+            guard let graphTarget = self.target(from: innerTarget),
+                  graphTarget.target.mergedBinaryType != .disabled
+            else {
+                continue
+            }
+
+            // Take all mergeable dependencies
+            let mergeableFrameworks = filterDependencies(
```

Comment:
I'd turn this one into a function and invoke it only in the case of `.automatic` and `.manual`. Otherwise we'll invoking it in the case of `.disabled`, even if we don't need it.

```swift
func mergeableFrameworks() -> [...] {
 ...
}
```

Reply:
Or even better, a private function in `GraphTraverser` to prevent adding more nesting to this function, which has already got too complex.